### PR TITLE
Handle data mapper codeaction failing scenarios

### PR DESCRIPTION
**Pull request by isuruRuhu**
## Purpose
Handle the data mapper code action failing scenarios. 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 1]
```diff
@@ -98,9 +113,9 @@
         String diagnosticMessage = diagnostic.message();
         Matcher matcher = CommandConstants.INCOMPATIBLE_TYPE_PATTERN.matcher(diagnosticMessage);
 
-        if (!(matcher.find() && matcher.groupCount() > 1)) {
-            return fEdits;
-        }
+//        if (!(matcher.find() && matcher.groupCount() > 1)) {
+//            return fEdits;
+//        }
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:118]: Shall we remove the commented code segment

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 2]
```diff
@@ -274,61 +303,210 @@ public static Symbol findSymbol(List<TypeSymbol> fileContentSymbols) {
...
         JsonObject rightRecordJSON = new JsonObject();
         JsonObject leftRecordJSON = new JsonObject();
 
-        // Schema 1
-        TypeSymbol rhsSymbol = ((TypeReferenceTypeSymbol) rhsTypeSymbol).typeDescriptor();
-        if ("RECORD".equals(rhsSymbol.typeKind().name())) {
-            RecordTypeSymbol rightSymbol = (RecordTypeSymbol) rhsSymbol;
-            Map<String, RecordFieldSymbol> rightSchemaFields = rightSymbol.fieldDescriptors();
+        List<RecordTypeSymbol> symbolList = checkMappingCapability(lftTypeSymbol, rhsTypeSymbol);
+        if (!symbolList.contains(null)) {
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:320]: Shall we avoid adding null entries to the list? Is there any specific reason/ limitation for adding the nulls

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 3]
```diff
@@ -274,61 +303,210 @@ public static Symbol findSymbol(List<TypeSymbol> fileContentSymbols) {
...
+
+
+            Map<String, RecordFieldSymbol> rightSchemaFields = symbolList.get(0).fieldDescriptors();
             JsonObject rightSchema = (JsonObject) recordToJSON(rightSchemaFields.values());
 
+            if (!restFieldMap.isEmpty()) {
+                rightSchema = insertRestFields(rightSchema, foundTypeRight);
+            }
+
+            if(!spreadFieldMap.isEmpty()){
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:338]: I think we don't need the empty check here.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 4]
```diff
@@ -75,11 +83,18 @@
     private static final int LEFT_SYMBOL_INDEX = 0;
     private static Cache<Integer, String> mappingCache =
             CacheBuilder.newBuilder().maximumSize(MAXIMUM_CACHE_SIZE).build();
+    private static HashMap<String, String> isOptionalMap = new HashMap<>();
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:86]: Why do we use static maps in a util? Is there a special reason for caching the older values and reuse them?

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 1 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java]: removed the segments

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 3 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java]: Some times spread fields can be empty.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 4 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:93]: I'm not caching older values. Just need to store them till generating the mapping function.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 2 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:394]: Fixed this. Stop adding null entries.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 5]
```diff
@@ -177,11 +234,17 @@
...
+                            matchedNode = ((TypeCastExpressionNode) positionDetails.matchedNode()).expression();
+                        }
                     }
 
                     if (matchedNode == null) {
-                        throw new IllegalStateException("Unexpected node at cursor:" +
-                                positionDetails.matchedNode().kind());
+                        return fEdits;
+//                        throw new IllegalStateException("Unexpected node at cursor:" +
+//                                positionDetails.matchedNode().kind());
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:247]: Remove the commented code segment

**Review by isuruRuhu — COMMENTED** [review]
Removed the commented lines

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    public void setModel(SemanticModel model) {
+        this.model = model;
+    }
+
+    @Override
+    public void visit(VariableDeclarationNode variableDeclarationNode) {
+        if (variableDeclarationNode.initializer().isPresent()) {
+            if (variableDeclarationNode.initializer().get().kind() == SyntaxKind.MAPPING_CONSTRUCTOR) {
+                String rightSymbolName = variableDeclarationNode.typedBindingPattern().typeDescriptor().toString().
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java:68]: Why do we use the string representation of the type descriptors and use `replaceAll` APIs here? We need to find an alternative approach to deal with this usage such as using the type descriptor's kind and etc

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                }
+                if (this.rhsSymbolName.equals(rightSymbolName)) {
+                    MappingConstructorExpressionNode mappingConstructorExpressionNode =
+                            (MappingConstructorExpressionNode) variableDeclarationNode.initializer().get();
+                    SeparatedNodeList<MappingFieldNode> fields = mappingConstructorExpressionNode.fields();
+                    for (MappingFieldNode field : fields) {
+                        if (field.kind() == SyntaxKind.SPECIFIC_FIELD) {
+                            SpecificFieldNode field1 = (SpecificFieldNode) field;
+                            String fieldName = field1.fieldName().toString().replaceAll(" ", "");
+                            if (field1.valueExpr().isPresent() && fieldName.contains("\"")) {
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java:84]: Can the field name contain `"`?

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java]: removed the lines

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        this.specificFieldList = new ArrayList<>();
+    }
+
+    public void setModel(SemanticModel model) {
+        this.model = model;
+    }
+
+    @Override
+    public void visit(VariableDeclarationNode variableDeclarationNode) {
+        if (variableDeclarationNode.initializer().isPresent()) {
```
[the remark this answers is not in the document]
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java:66]: Modified the code according to the comment.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            if (variableDeclarationNode.initializer().get().kind() == SyntaxKind.MAPPING_CONSTRUCTOR) {
```
[the remark this answers is not in the document]
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java:67]: Modified the code according to the comment.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java (hunk 7 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java:84]: When the field is a rest field, there are `"` in the field name.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java (hunk 6 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java]: Modified the code according to the comments

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    public void setModel(SemanticModel model) {
+        this.model = model;
+    }
+
+    @Override
+    public void visit(VariableDeclarationNode variableDeclarationNode) {
+        if (!(variableDeclarationNode.initializer().isPresent())) {
+            return;
+        }
+        if (!(variableDeclarationNode.initializer().get().kind() == SyntaxKind.MAPPING_CONSTRUCTOR)) {
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java:72]: use` !=`

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        this.specificFieldList = new ArrayList<>();
+    }
+
+    public void setModel(SemanticModel model) {
+        this.model = model;
+    }
+
+    @Override
+    public void visit(VariableDeclarationNode variableDeclarationNode) {
+        if (!(variableDeclarationNode.initializer().isPresent())) {
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java:69]: use `isEmpty`

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java (hunk 10 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java:72]: Modified the code accordingly

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java (hunk 11 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/DataMapperNodeVisitor.java:69]: Modified the code accordingly.

**Review by isuruRuhu — COMMENTED** [review]
Modified the code according to the comments

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 12]
```diff
@@ -130,8 +165,37 @@
...
+                        rhsTypeSymbol = (Symbol) props.get(LEFT_SYMBOL_INDEX).value();
+
+                        foundTypeLeft = lftTypeSymbol.getName().orElse("");
+                        foundTypeRight = rhsTypeSymbol.getName().orElse("");
+                    } else {
+                        return fEdits;
+                    }
+                } else {
+                    return fEdits;
+                }
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:197]: You can use the early return and reduce the nested if

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 13]
```diff
@@ -196,6 +271,22 @@
...
+                        String checkExpression = "";
+                        Collection<ChildNodeEntry> childEntries = matchedNode.childEntries();
+                        for (ChildNodeEntry childEntry : childEntries) {
+                            if (childEntry.name().equals("expression")) {
+                                functionCall = childEntry.node().isPresent() ?
+                                        childEntry.node().get().toString() : null;
+                            } else if (childEntry.name().equals("expression")) {
+                                checkExpression = childEntry.node().isPresent() ?
+                                        childEntry.node().get().toString() : null;
+                            }
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:284]: Why do we use the same expression here? Only the if clause is executed. Revamp this logic

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeAction.java [hunk 14]
```diff
@@ -42,7 +42,7 @@
 import java.util.Locale;
 import java.util.Optional;
 
-import static org.ballerinalang.datamapper.AIDataMapperCodeActionUtil.getAIDataMapperCodeActionEdits;
+//import static org.ballerinalang.datamapper.AIDataMapperCodeActionUtil.getAIDataMapperCodeActionEdits;
```
[the remark this answers is not in the document]
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeAction.java:45]: Removed the line

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 15]
```diff
@@ -73,14 +87,40 @@
...
 
     private static final String SCHEMA = "schema";
     private static final String ID = "id";
     private static final String TYPE = "type";
     private static final String PROPERTIES = "properties";
+    private static final String OPTIONAL = "optional";
+    private static final String READONLY = "readonly";
+    private static final String SIGNATURE = "signature";
+
+    //static member holds only one instance of the JDBCSingleton class.
```
[the remark this answers is not in the document]
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:110]: Removed the lines

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 12 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:197]: modified the code according to the comment

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 13 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java]: changed the logic

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 16]
```diff
@@ -177,11 +246,15 @@
                         if (matchedNode.parent().kind() == SyntaxKind.CHECK_EXPRESSION) {
                             matchedNode = matchedNode.parent();
                         }
+                    } else if (positionDetails.matchedNode().kind() == SyntaxKind.TYPE_CAST_EXPRESSION) {
+                        if (((TypeCastExpressionNode) positionDetails.matchedNode()).expression().kind() ==
+                                SyntaxKind.FUNCTION_CALL) {
+                            matchedNode = ((TypeCastExpressionNode) positionDetails.matchedNode()).expression();
+                        }
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:253]: We can merge these two conditions together

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 17]
```diff
@@ -274,81 +383,294 @@ public static Symbol findSymbol(List<TypeSymbol> fileContentSymbols) {
...
+        return getMapping(schemas, context);
+    }
+
+    private JsonObject insertRestFields(JsonObject rightSchema, String foundTypeRight) {
+        // Added to get the rest fields
+        HashMap<String, String> tempRestFieldMap = new HashMap<>(this.restFieldMap);
+        Set<Map.Entry<String, String>> entrySet = tempRestFieldMap.entrySet();
+        for (Map.Entry<String, String> restField : entrySet) {
+            JsonObject fieldDetails = new JsonObject();
+            fieldDetails.addProperty(ID, "dummy_id");
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:468]: Avoid using the `dummy_` prefix and use something else

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 17 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:468]: removed `dummy_`

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 16 above)
Comment (isuruRuhu) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java]: Merged the two conditions

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Review by malithjayasinghe — APPROVED** [review]

**Pull request merged**